datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -1532,17 +1532,21 @@ fn get_filtered_join_mask(
             for i in 0..streamed_indices_length {
                 // LeftSemi respects only first true values for specific streaming index,
                 // others true values for the same index must be false
-                if mask.value(i) && !seen_as_true {
+                let streamed_idx = streamed_indices.value(i);
+                if mask.value(i)
+                    && !seen_as_true
+                    && !matched_indices.contains(&streamed_idx)
```

Comment:
this is an actual fix!

Reply:
Hmm, the fix is correct. I roughly remember I have commented on such case. `JoinType::LeftAnti` below has considered `matched_indices`. Maybe for `JoinType::LeftSem` it is missing.